pkg/kernel/stage.go

```diff
@@ -76,61 +87,81 @@ func (s *stage) run(k *kernel) {
 	}
 
 	s.running.Signal()
+
+	return s.waitUntilHealthy()
 }
 
-func (s *stage) stopWait(stageIndex int, logger log.Logger) {
-	s.cfn.Kill(ErrKernelStopping)
-	s.err = s.cfn.Wait()
+func (s *stage) healthcheck() HealthCheckResult {
+	var ok bool
+	var err error
+	var healthAware HealthCheckedModule
+	var result HealthCheckResult
+
+	for name, ms := range s.modules.modules {
+		if healthAware, ok = ms.module.(HealthCheckedModule); !ok {
+			continue
+		}
 
-	if s.err != nil && s.err != ErrKernelStopping {
-		logger.Error("error during the execution of stage %d: %w", stageIndex, s.err)
+		ok, err = healthAware.IsHealthy(s.ctx)
+
+		result = append(result, ModuleHealthCheckResult{
+			StageIndex: s.index,
+			Name:       name,
+			Healthy:    ok,
+			Err:        err,
+		})
 	}
 
-	s.terminated.Signal()
+	return result
 }
 
-func (s *stage) len() int {
-	return s.modules.len()
-}
+func (s *stage) waitUntilHealthy() error {
+	var result HealthCheckResult
+
+	timeoutTicker := clock.NewRealTimer(s.healthCheckSettings.Timeout)
+	sleepTicker := clock.NewRealTicker(s.healthCheckSettings.WaitInterval)
 
-type stages map[int]*stage
+	for {
+		sleepTicker.Stop()
+		result = s.healthcheck()
 
-func (s stages) hasModules() bool {
-	// no need to iterate in order as we are only checking
-	for _, stage := range s {
-		if len(stage.modules.modules) > 0 {
-			return true
+		if result.IsHealthy() {
+			return result.Err()
 		}
-	}
 
-	return false
-}
+		if result.Err() != nil {
+			s.logger.Warn("errors during health checks in stage %d: %s", s.index, result.Err())
```

Comment:
Do we log the errors on the health check side? I assume it's more informative with the context around there

Reply:
I've added logging to the api health checks and also added the context there: `pkg/apiserver/health_check.go:89`  
Is that answering / solving your question?

Reply:
Yes